src/signatures.rs

```diff
@@ -0,0 +1,1206 @@
+use concordium_base::{
+    common::{
+        types::{CredentialIndex, KeyIndex, KeyPair, Signature},
+        Versioned,
+    },
+    contracts_common::{AccountAddress, SignatureThreshold},
+    curve_arithmetic::Curve,
+    id::types::{
+        AccountCredentialWithoutProofs, AccountKeys, Attribute, InitialAccountData,
+        PublicCredentialData, VerifyKey,
+    },
+};
+use ed25519_dalek::{SigningKey, VerifyingKey};
+use sha2::Digest;
+use std::collections::BTreeMap;
+
+/// Errors that can occur during generating or verifying account signatures.
+#[derive(thiserror::Error, Debug)]
+pub enum SignatureError {
+    #[error("Network error: {0}")]
+    QueryError(#[from] QueryError),
+    #[error(
+        "Indices do not exist on chain for credential index `{credential_index}` and key index \
+         `{key_index}`"
+    )]
+    MissingIndicesOnChain {
+        credential_index: u8,
+        key_index:        u8,
+    },
+    #[error("The indices in the maps do not match")]
+    MismatchMapIndices,
+    #[error(
+        "The public key and the private key in the `account_keys` map do not match for credential \
+         index `{credential_index}` and key index `{key_index}`"
+    )]
+    MismatchPublicPrivateKeys {
+        credential_index: u8,
+        key_index:        u8,
+    },
+    #[error(
+        "The public key on chain `{expected_public_key:?}` and the public key \
+         `{actual_public_key:?}` in the signature map do not match for credential index \
+         `{credential_index}` and key index `{key_index}`"
+    )]
+    MismatchPublicKeyOnChain {
+        credential_index:    u8,
+        key_index:           u8,
+        expected_public_key: Box<VerifyingKey>,
+        actual_public_key:   Box<VerifyingKey>,
+    },
+}
+
+/// Message types that can be signed and verified. The Concordium wallets can
+/// either sign binary or text messages.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Message<'a> {
+    BinaryMessage(&'a [u8]),
+    TextMessage(&'a str),
+}
+
+/// Account signatures are constructed similarly to transaction signatures. The
+/// only difference is that account signatures use as account nonce the value 0
+/// when signing. This is to ensure that the user does not accidentally sign a
+/// valid transaction. As such account signatures can be used to sign messages
+/// in wallets that can then be verified in a smart contract or by a backend.
+///
+/// Account signatures should be thought of as a nested map, indexed on the
+/// outer layer by credential indices, and the inner map maps key indices to
+/// [`Signature`]s.
+#[derive(Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct AccountSignatures {
+    pub sigs: BTreeMap<u8, CredentialSignatures>,
+}
+
+#[derive(Debug, PartialEq, Eq, serde::Deserialize, serde::Serialize)]
+#[repr(transparent)]
+pub struct CredentialSignatures {
+    pub sigs: BTreeMap<u8, Signature>,
+}
```

Comment:
It would influence the serde implementation though

Reply:
It would yes, but the `repr(transparent)` kinda suggests that it should be treated as just the btree. But yea dunno if the serialized version's layout is significant here.

Reply:
Currently, the serde implementation seems not relevant but could become depending on what we add in the future or use this type for.

I agree with that the `unit` types would be nicer but since we use these types of 2-level maps already in our code base whenever dealing with identities we might just stay consistent with it. Will keep the current implementation.
https://github.com/Concordium/concordium-base/blob/main/smart-contracts/contracts-common/concordium-contracts-common/src/types.rs#L698

Reply:
Just be aware that the pub field also allows the user to do whatever they want with the map, including making it empty for example. The code using the type should be okay with that.